### DISCUSSION TITLE
Is a difference between Arcore pose.translation and Filament Vertex?

### DISCUSSION THREAD

**Discussion by inveders** [google/filament#4408] [Q&A] [answered]
I want to create a shape with Vertex from arCore, I retrieve, by my hitTest, my translation vector from arCore and transform it with Vector3 to be in world space. But when I put this values to create my shape with Vertex, it's not seems good. The position of my hitTest don't match with the position of the same point draw with FIlament.
Is there a difference, or need a conversion?

**Comment by romainguy** [comment, marked as answer]
If I remember correctly, ARCore's hit results return a `Pose` that's always in world space. You shouldn't need to convert it.

**Reply by inveders** [reply]
Yes that it, sorry it was not a conversion @romainguy , I just put it in a Vector3 to be more manipulable (see more easily the x, y and z). But when I put theses points in my Vertex they don't draw well. I don't know if there is something to do about it?
For example, on the next image, I have my rectangle drawing dynamically, but the little white point is really where I have my HitResult Point. So I retrieve the coordinates of this little white point, and it would be the coordinates in my Vertex for draw one part of the rectangle. But, as you see, the rectangle is not on the little white point.
<img width="408" alt="Capture d’écran 2021-07-29 à 18 39 39" src="https://user-images.githubusercontent.com/40520645/127531243-1bf7181c-3ac5-4971-a646-1af5406a2c0e.png">

**Reply by romainguy** [reply]
You might be missing a transform on that point? With ARCore there's a transform that needs to be applied to the entire scene I believe for things to work properly (`Camera.getDisplayOrientedPose()` I think).

**Reply by inveders** [reply]
Yes I have already apply this Camera.getDisplayOrientedPose() on my scene. It's just weird that with the same coordinates (setTransform for the little white point and Vertex for the Rectangle) I don't retrieve the same position on the screen... I don't know :-/

**Reply by inveders** [reply]
@romainguy Is it possible that there is a problem between ArCore's getProjectionMatrix and Filament's understanding of it? Because I have two points retrieved from the HitResult on a vertical surface, and when I draw my rectangle with FIlament, I also have the same problem, my rectangle is much too small compared to the two points I have, as if the z-axis point I retrieved from my points during the hitResult is not the right z that is drawn. Do I need to do a conversion between the point I retrieve and the point I draw?

And **second question** : on wich point the setTransform is for my rectangle?
For example I have
```kotlin
engine.transformManager.setTransform(
            engine.transformManager.getInstance(renderable),
            m4Identity()
                .translate(mCPoint.x, mAPoint.y, mAPoint.z) //Correspond to the D Point
                .rotate(0f.toDegrees, 0f, 1f, 0f)
                .scale(1f, 1f, 1f)
                .floatArray,
        )
```
and my rectangle vertexData is
```kotlin
return ByteBuffer.allocate(vertexCount * 3 * Float.SIZE_BYTES)
                            // It is important to respect the native byte order
                            .order(ByteOrder.nativeOrder())
                            .put(Vertex(mAPoint.x, mAPoint.y, mAPoint.z)) //A point with A on left high position
                            .put(Vertex(mAPoint.x, mCPoint.y, mAPoint.z)) //B point
                            .put(Vertex(mCPoint.x, mCPoint.y, mAPoint.z)) // C Point
                            .put(Vertex(mCPoint.x, mAPoint.y, mAPoint.z)) // D Point
                            // Make sure the cursor is pointing in the right place in the byte buffer
                            .flip()
```
Where is the setTransform? On the D Point? So My rectangle is put on the position of the setTransform from the D Point? Is right?

**Reply by romainguy** [reply]
> Where is the setTransform? On the D Point? So My rectangle is put on the position of the setTransform from the D Point? Is right?

I don't understand your question. What are you trying to do?

**Reply by inveders** [reply]
@romainguy I'm trying to understand how I set the position of my rectangle according to the setTransform. I try to draw a rectangle with two points, A and C (with A the high left vertice). But when I finish to select my points, the rectangle is like drawn too small and his position start in the high left. I need to understand on wich vertices of the rectangle the setTransform is based to place my rectangle in the space according to the vertex data?And why my rectangle is not drawn until the C point. I don't know if it's more understandable?
https://drive.google.com/file/d/17EPTdzPStbCePqwf_QRoIi9B9fGysyFg/view?usp=drivesdk

**Reply by romainguy** [reply]
> on wich vertices of the rectangle the setTransform is based to place my rectangle

`setTransform()` is not based on a specific vertex, it applies to all the vertices. Anyway think of this way:

- The vertices you put in the vertex buffer are in object space, a local coordinates system that's specific to that object.
- The transform you set on a renderable with `setTransform()` will transform each vertex of the renderable from object space to world space (a coordinates system common to all the objects in the scene).
- The (inverse of the) transform set on the camera will be used to transform each vertex from world space to view space (a coordinates system common to all the objects in the scene by relative to the camera)

Again, I don't quite understand what you are trying to do but you should define your rectangle in an object space that's easy to manipulate. For instance, always use 0,0 for one of the corners of the rectangle and use the width and the height you want to define the other points. Then you can set a transform on your rectangle to position it where you want in the 3D environment.

>  But when I finish to select my points, the rectangle is like drawn too small and his position start in the high left

What are those two points? Where do they come from? Are they in view space, world space, something else?

**Reply by inveders** [reply]
@romainguy Ok thanks you I understand better the way the shapes are drawn in this system. 

The points are coming from the hitResult on my vertical plane detected with arCore (I tap with my finger on two places like in image in the link above and a rectangle is drawn from this two points). So I need to draw a rectangle on my vertical planes from a specific place I choose (where I tap). To summarize I want to set my rectangle to match a specific point of my rectangle (the A high left point) in a specific place. I ill try to do the way you define above.

**Reply by romainguy** [reply]
You should check the documentation of ARCore to know what space those points are in exactly. I assume they are in world space for a given frame. Note that a world space coordinate in ARCore is only valid for the *current* frame unless you use an `Anchor`. See "World Coordinate Space" in the [Pose documentation](https://developers.google.com/ar/reference/java/com/google/ar/core/Pose). It sounds like you'll need to create anchors to achieve what you want.
